### PR TITLE
TOOLS-249 Implement transaction search by querying the DB

### PR DESCRIPTION
**Pull request by mpaddle**
## Ticket(s):
Link to ticket:
https://paddle.atlassian.net/browse/TOOLS-249 

## Checklist
- [x] I have added tests or left a comment on why they are not needed
- [x] I deployed and tested the PR on development

## What has been done:

1. I've written and annotated a number of queries based on previous experience to retrieve transaction data based on the given input.
1. The queries have been converted to Python and integrated with the empty Lambda.
1. The Lambda has been tested on development and staging. So far the response appears to be correct and rarely timing out. Therefore something must be wrong with the code.

## How to test:

There are a great number of scenarios to test. Most of this comes down to the "state" value. I've tried to describe below the different entities and their states.

The base entities are:

* One-off purchase (non-subscription products)
* Initial subscription payment (when you purchase a subscription)
* Subsequent subscription payment (the renewal of a subscription) 

For each base entity, there are a number of states:

* Complete: no refunds or all refunds have been deleted, and no chargeback or all chargebacks have been deleted.
* Partially refunded: 1 or more non-declined refunds that do not amount to the total amount of the transaction, and no chargeback or all chargebacks deleted.
* Refunded: same as partially refunded but now the non-declined, non-deleted refund total adds up to the amount of the transaction.
* Disputed: 1 non-deleted chargeback for the transaction; refunds are ignored as chargebacks take precedence.

### Testing instructions

1. Authenticate using Okta-STS to call the Lambda in development or staging. Optionally you can call the Lambda on the AWS dashboard.
1. Find a relevant transaction in the staging database (or through the seller dashboard).
1. Enter the relevant details of your transaction in the Lambda call below:
    ```
    aws lambda invoke --profile=staging \
        --function-name=paddle-staging-transaction-search-fetch \
        --payload='{"last_four_digits":"<digits>","date_from":"yyyy-mm-dd","date_to":"yyyy-mm-dd","amount":"<amount>"}' test.txt \
        && cat test.txt | jq .
    ```
    If you are using the AWS dashboard, enter the payload directly: `{"last_four_digits":"<digits>","date_from":"yyyy-mm-dd","date_to":"yyyy-mm-dd","amount":"<amount>"}`.
1. Confirm that the Lambda returns your transaction. Note that on staging we use test cards which typically all have the same card number (e.g. 4111-1111-1111-1111). So you may have multiple, unrelated transactions in your response.

## How to deploy to production:

1. Additional deployment permissions have been added: create or update the CloudFormation stack for the deployment user.
1. New SSM keys are needed: add the following SSM keys to production with values retrieved from appropriate sources. Remember to use the _readonly_ database!
    * `/transaction-search/staging/monolith-db/user`
    * `/transaction-search/staging/monolith-db/password`
    * `/transaction-search/staging/monolith-db/host`
    * `/transaction-search/staging/monolith-db/name`

### CODE REVIEW

File: src/transaction_search/fetch.py [hunk 1]
```diff
@@ -1,8 +1,27 @@
...
 def handler(event: Dict[str, Any], context: LambdaContext) -> Dict:
-    return {"statusCode": 200, "body": json.dumps({"message": "All good!"})}
+    if "last_four_digits" not in event:
+        return {"statusCode": 400, "errors": [{"title": "missing last_four_digits"}]}
+    if "date_from" not in event:
+        return {"statusCode": 400, "errors": [{"title": "missing date_from"}]}
+    if "date_to" not in event:
+        return {"statusCode": 400, "errors": [{"title": "missing date_to"}]}
+    if "amount" not in event:
+        return {"statusCode": 400, "errors": [{"title": "missing amount"}]}
```
Comment (vhemt) [inline src/transaction_search/fetch.py]: 💡 💡 💡 (three times cause it's just a matter of feeling cool 😎 ):
returning a list of errors based on the missing parameters?

Comment (mpaddle) [reply src/transaction_search/fetch.py]: Can do 👍  At the time I did the simplest thing but it won't take long to change

File: tests/transaction_search/test_fetch.py [hunk 2]
```diff
@@ -1,7 +1,6 @@
-from transaction_search.fetch import handler
+# from transaction_search.fetch import handler
```
Comment (vhemt) [inline tests/transaction_search/test_fetch.py]: ^

File: payloads/Fetch.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+    "last_four_digits": "1111",
+    "date_from": "2019-09-01",
+    "date_to": "2019-09-05",
+    "amount": "19.50"
+}
```
Comment (vhemt) [inline payloads/Fetch.json]: new line

File: tests/transaction_search/test_fetch.py (hunk 2 above)
Comment (mpaddle) [reply tests/transaction_search/test_fetch.py]: Life's too short for tests 🤷‍♂

See Slack for the ongoing conversation about testing this temporary feature

Comment (vhemt) [reply tests/transaction_search/test_fetch.py]: oh agree on that, just remove the commented line 😄

**Comment by vhemt** [comment]
looks good to me, I need to test it though. Will do in a bit 👍

File: tests/transaction_search/test_fetch.py (hunk 2 above)
Comment (mpaddle) [reply tests/transaction_search/test_fetch.py]: Fine, as long as I'm not getting told off for not having tests 😂

File: src/transaction_search/fetch.py (hunk 1 above)
Comment (mpaddle) [reply src/transaction_search/fetch.py]: Done

File: tests/transaction_search/test_fetch.py (hunk 2 above)
Comment (mpaddle) [reply tests/transaction_search/test_fetch.py]: Done

File: payloads/Fetch.json (hunk 3 above)
Comment (mpaddle) [reply payloads/Fetch.json]: Done

File: src/transaction_search/fetch.py (hunk 1 above)
Comment (vhemt) [reply src/transaction_search/fetch.py]: just realised something, are the parameters optional maybe?
Do we actually want to search using all 4 parameters?

Comment (mpaddle) [reply src/transaction_search/fetch.py]: I think we need all 4: any less than these parameters and you'll have too many results. But we can confirm with Greg.

**Comment by vhemt** [comment]
works on staging, I didn't checked if all the data is valid directly from the DB, but seems 👌

**Pull request merged**